dali/operators/generic/slice/subscript.h

```diff
@@ -0,0 +1,325 @@
+    step.Init(spec, make_string("step_", i));
+
+    if (at.IsDefined()) {
+      if (lo.IsDefined() || hi.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified both as an index and as a range."));
+      }
+      if (step.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified as an index - it cannot have a step."));
+      }
+    }
+  }
+
+  bool IsDefined() const {
+    return at.IsDefined() || lo.IsDefined() || hi.IsDefined() || step.IsDefined();
+  }
+
+  bool IsIndex() const {
+    return at.IsDefined();
+  }
+
+  bool IsRange() const {
+    return !IsIndex() && IsDefined();
+  }
+
+  SubscriptArg at, lo, hi, step;
+
+  void Load(const ArgumentWorkspace &ws, int batch_size) {
+    at.Load(ws, batch_size);
+    lo.Load(ws, batch_size);
+    hi.Load(ws, batch_size);
+    step.Load(ws, batch_size);
+  }
+};
+
+
+template <typename Backend>
+class TensorSubscript : public Operator<Backend> {
+ public:
+  explicit TensorSubscript(const OpSpec &spec) : Operator<Backend>(spec) {
+    InitArgs();
+  }
+
+  USE_OPERATOR_MEMBERS();
+
+  void InitArgs() {
+    subscripts_.resize(kMaxSubscripts);
+    int num_subscripts = 0;
+    for (int i = 0; i < kMaxSubscripts; i++) {
+      SubscriptInfo &s = subscripts_[i];
+      s.Init(spec_, i);
+
+
+      if (s.IsDefined()) {
+        num_subscripts = i + 1;
+      }
+    }
+    subscripts_.resize(num_subscripts);
+  }
+
+  bool SetupImpl(vector<OutputDesc> &outputs, const workspace_t<Backend> &ws) override {
+    const auto &input = ws.template InputRef<Backend>(0);
+    DALI_ENFORCE(input.sample_dim() > 0, "Cannot apply an index to a scalar.");
+    const auto &input_shape = input.shape();
+    outputs.resize(1);
+    outputs[0].type = input.type();
+    GetRanges(ws, input_shape);
+    ProcessShapes(outputs[0].shape, input_shape);
+    return true;
+  }
+
+  void GetRanges(const workspace_t<Backend> &ws, const TensorListShape<> &in_shape) {
+    int nsub = subscripts_.size();
+    int ndim = in_shape.sample_dim();
+    DALI_ENFORCE(ndim >= nsub, "Too many indices.");
+    int nsamples = in_shape.num_samples();
+    lo_.resize(nsamples, ndim);
+    hi_ = in_shape;
+    step_.resize(nsamples, ndim);
+    for (auto &s : step_.shapes)
+      s = 1;
+    shape_ = in_shape;
+
+    for (int d = 0; d < nsub; d++) {
+      SubscriptInfo &s = subscripts_[d];
+      s.Load(ws, nsamples);
+      GetRange(s, d, in_shape);
+    }
+  }
+
+  void GetRange(SubscriptInfo &s, int d, const TensorListShape<> &in_shape) {
+    int nsamples = in_shape.num_samples();
+
+    if (s.IsIndex()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t at = s.at.values[i];
+        int64_t idx = at < 0 ? in_extent + at : at;
+        DALI_ENFORCE(idx >= 0 && idx < in_extent, make_string("Index ", idx, " is out of range "
```

Comment:
```suggestion
        DALI_ENFORCE(idx >= 0 && idx < in_extent, make_string("Index ", idx, "(", at, ") is out of range "
```

Reply:
I'll just use `at` here - using two values for positive indices would look weird and for negative indices we could assume that the users know what they're doing.

Reply:
...and I'll replace `DALI_ENFORCE` with plain `if` and `DALI_FAIL` to avoid putting the confusing condition in the error message.